src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,467 @@
+	                                              spin,
+	                                              hdd_seek_samples,
+	                                              true);
+	active_devices.push_back(hdd_noise);
+
+	floppy_noise = std::make_shared<DiskNoiseDevice>(DiskType::Floppy,
+	                                                 enable_floppy_disk_noise,
+	                                                 floppy_spin_up,
+	                                                 floppy_spin,
+	                                                 floppy_seek_samples,
+	                                                 false);
+	active_devices.push_back(floppy_noise);
+	MIXER_UnlockMixerThread();
+}
+
+void DiskNoises::AudioCallback(const int num_frames_requested)
+{
+	// stereo interleaved buffer
+	static std::vector<AudioFrame> out = {};
+	out.clear();
+
+	// Mix audio frames from all active devices
+	for (int i = 0; i < num_frames_requested; ++i) {
+		AudioFrame mixed_sample = {0.0f, 0.0f};
+		for (const auto& device : disk_noises->active_devices) {
+			AudioFrame sample = device->GetNextFrame();
+			mixed_sample += sample;
+		}
+		out.emplace_back(mixed_sample);
+	}
+
+	disk_noises->mix_channel->AddAudioFrames(out);
+}
+
+DiskNoises::~DiskNoises()
+{
+	MIXER_LockMixerThread();
+
+	if (floppy_noise) {
+		floppy_noise.reset();
+	}
+
+	if (hdd_noise) {
+		hdd_noise.reset();
+	}
+	active_devices.clear();
+
+	if (active_devices.empty() && mix_channel) {
+		mix_channel->Enable(false);
+		MIXER_DeregisterChannel(mix_channel);
+		mix_channel.reset();
+	}
+
+	MIXER_UnlockMixerThread();
+}
+
+AudioFrame DiskNoiseDevice::GetNextFrame()
+{
+	const float DiskNoiseGain = 0.2f;
+	float sample              = 0.0f;
+
+	// Mix in spinup and spin samples
+	if (!spin.spin_up_sample.empty() &&
+	    spin.spin_up_it != spin.spin_up_sample.end()) {
+		sample += (*spin.spin_up_it) * DiskNoiseGain;
+		++spin.spin_up_it;
+	} else if (!spin.sample.empty() &&
+	           (spin.spin_it != spin.sample.end() || spin.loop)) {
+		// Loop the spin sound if enabled. Used for
+		// persistent HDD noise Not used for floppy
+		// noise because motor should stop after r/w
+		// operations aredone
+		if (spin.spin_it == spin.sample.end() && spin.loop) {
+			spin.spin_it = spin.sample.begin();
+		}
+		if (spin.spin_it != spin.sample.end()) {
+			sample += (*spin.spin_it) * DiskNoiseGain;
+			++spin.spin_it;
+		}
+	}
+
+	// Mix in seek sample, if it's playing
+	if (!seek.current_sample.empty() &&
+	    seek.current_it != seek.current_sample.end()) {
+		sample += (*seek.current_it) * DiskNoiseGain;
+		++seek.current_it;
+	}
+
+	// If seek sample has finished playing, clear entry
+	if (!seek.current_sample.empty() &&
+	    seek.current_it == seek.current_sample.end()) {
+		seek.current_sample.clear();
+		seek.current_it = seek.current_sample.end();
+	}
+
+	return AudioFrame{sample, sample};
+}
+
+void DiskNoiseDevice::LoadSample(const std::string& path, std::vector<float>& buffer)
```

Comment:
`dest_buffer` or `buffer_out` would be better. We should make when a function argument is an "out" arg.